src/compiler/core.ts

```diff
@@ -92,6 +92,41 @@ module ts {
         return result;
     }
 
+    export function reduce<T>(array: T[], f: (a: T, x: T) => T): T;
+    export function reduce<T, U>(array: T[], f: (a: U, x: T) => U, initial: U): U;
+    export function reduce<T, U>(array: T[], f: (a: any, x: T) => any, initial?: any): any {

```

Comment:
@JsonFreeman but then you'd contextually type with `{}` when `initial` is not set.